storybookjs/storybook
# viewport-addon: Make the addon configurable

Note: This is my first open-source PR.

### Edit
I addressed #2395 and the respective [comment](https://github.com/storybooks/storybook/issues/2395#issuecomment-369578940)

Issue: Currently you cannot customize or add new viewports/devices as described in the README file

## What I did
* Exported a `configure` function to easily customize the addon
  * Add new device/viewport
  * Replace the entire viewports set
   * Change the default viewport for all stories
* Exported a decorator function and wrapper component to change the default viewport per story
* Replaced the `default` option with a new viewport called `responsive` to treat the default as 1st-class citizen
* Added 2 new devices, and different stories for the`viewport` addon to the `official-storybook` example
* Organized the code respectively

### Add new device
```javascript
// config.js
import { configure, INITIAL_VIEWPORTS } from '@storybook/addon-viewport';

const extraViewports = {
  kindleFire2: {
    name: 'Kindle Fire 2',
    styles: {
      width: '600px',
      height: '963px'
    }
  },
  kindleFireHD: {
    name: 'Kindle Fire HD',
    styles: {
      width: '533px',
      height: '801px'
    }
  }
};

configure({
  viewports: {
    ...INITIAL_VIEWPORTS,
    ...extraViewports
  }
});
```
<img width="795" alt="add" src="https://user-images.githubusercontent.com/3688136/37190662-5c31672a-235b-11e8-88ef-d07f369dddc6.png">


### Use new list of devices
```javascript
// config.js

import { configure } from '@storybook/addon-viewport';

const newViewports = {
  kindleFire2: {
    name: 'Kindle Fire 2',
    styles: {
      width: '600px',
      height: '963px'
    }
  },
  kindleFireHD: {
    name: 'Kindle Fire HD',
    styles: {
      width: '533px',
      height: '801px'
    }
  }
};

configure({
  viewports: newViewports
});
```

<img width="795" alt="replace" src="https://user-images.githubusercontent.com/3688136/37260255-29747cb8-2591-11e8-870e-79c602c7d58e.png">


### Change the default viewport

```javascript
// config.js

import { configure } from '@storybook/addon-viewport';

configure({
  defaultViewport: 'iphone6'
});
```

<img width="393" alt="default-for-all" src="https://user-images.githubusercontent.com/3688136/37190879-b84d307e-235c-11e8-9d91-93bf483d9b94.png">


### Change the default viewport per story

```javascript
// file.stories.js

import React from 'react';
import { storiesOf } from '@storybook/react';
import { withViewport, Viewport } from '@storybook/addon-viewport';

storiesOf('Kindle Fire 2', module)
  .addDecorator(withViewport('kindleFire2'))
  .add('Inherited', () => (
    <div>
      I will be displayed on <b>Kindle Fire 2</b> viewport by default.
    </div>
  ))
  .add('Overridden', () => (
    <Viewport name="iphone6">
      <div>
        I will be displayed on <b>iPhone 6</b> viewport by default.
      </div>
    </Viewport>
  ));
```

## How to test
1. `cd examples/official-storybook`
2. `yarn storybook`
3. You will see 2 new devices were added to the `viewport` panel
4. Change the `extra-viewports.json` file
5. Observe the changes

Is this testable with jest or storyshots?
Yes

Does this need a new example in the kitchen sink apps?
No, I added it to the `official-storybook` already

Does this need an update to the documentation?
Yes

If your answer is yes to any of these, please make sure to include it in your PR.
